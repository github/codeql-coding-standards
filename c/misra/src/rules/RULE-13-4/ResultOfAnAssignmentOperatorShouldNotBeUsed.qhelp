<!DOCTYPE qhelp PUBLIC "-//Semmle//qhelp//EN" "qhelp.dtd">
<qhelp>
  <!-- THIS FILE IS AUTOGENERATED, DO NOT MODIFY DIRECTLY. -->
  <overview>
    <p>This query implements the MISRA-C-2012 rule RULE-13-4:</p>
    <blockquote>
      <p>The result of an assignment operator should not be used</p>
    </blockquote>
  </overview>
  <include src="ResultOfAnAssignmentOperatorShouldNotBeUsed-standard.qhelp" />
  <section title="Implementation notes">
    <include src="ResultOfAnAssignmentOperatorShouldNotBeUsed-implementation.qhelp" />
  </section>
  <references>
    <li>
      MISRA-C:2012 Guidelines for the use of the C language in critical systems:
      <a href="https://www.misra.org.uk/">RULE-13-4: The result of an assignment operator should not be used</a>
      .
    </li>
  </references>
</qhelp>