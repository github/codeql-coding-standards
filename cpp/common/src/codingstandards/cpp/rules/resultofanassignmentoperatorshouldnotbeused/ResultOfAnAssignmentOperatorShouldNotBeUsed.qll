/**
 * Provides a library with a `problems` predicate for the following issue:
 * The result of an assignment operator should not be used.
 */

import cpp
import codingstandards.cpp.Customizations
import codingstandards.cpp.Exclusions

abstract class ResultOfAnAssignmentOperatorShouldNotBeUsedSharedQuery extends Query { }

Query getQuery() { result instanceof ResultOfAnAssignmentOperatorShouldNotBeUsedSharedQuery }

query predicate problems(AssignExpr e, string message) {
  not isExcluded(e, getQuery()) and
  not exists(ExprStmt s | s.getExpr() = e) and
  message = "Use of an assignment operator's result."
}
