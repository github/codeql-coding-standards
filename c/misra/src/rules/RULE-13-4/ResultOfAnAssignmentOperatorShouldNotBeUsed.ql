/**
 * @id c/misra/result-of-an-assignment-operator-should-not-be-used
 * @name RULE-13-4: The result of an assignment operator should not be used
 * @description The use of an assignment operator can impair the readability of the code and the
 *              introduced side effect may result in undefined behavior.
 * @kind problem
 * @precision very-high
 * @problem.severity error
 * @tags external/misra/id/rule-13-4
 *       correctness
 *       readability
 *       external/misra/obligation/advisory
 */

import cpp
import codingstandards.c.misra
import codingstandards.cpp.rules.resultofanassignmentoperatorshouldnotbeused.ResultOfAnAssignmentOperatorShouldNotBeUsed

class ResultOfAnAssignmentOperatorShouldNotBeUsedQuery extends ResultOfAnAssignmentOperatorShouldNotBeUsedSharedQuery
{
  ResultOfAnAssignmentOperatorShouldNotBeUsedQuery() {
    this = SideEffects1Package::resultOfAnAssignmentOperatorShouldNotBeUsedQuery()
  }
}
