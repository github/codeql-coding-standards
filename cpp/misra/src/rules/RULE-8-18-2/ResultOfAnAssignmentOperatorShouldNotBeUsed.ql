/**
 * @id cpp/misra/result-of-an-assignment-operator-should-not-be-used
 * @name RULE-8-18-2: The result of an assignment operator should not be used
 * @description The result of an assignment operator should not be used.
 * @kind problem
 * @precision very-high
 * @problem.severity error
 * @tags external/misra/id/rule-8-18-2
 *       scope/single-translation-unit
 *       external/misra/enforcement/decidable
 *       external/misra/obligation/advisory
 */

import cpp
import codingstandards.cpp.misra
import codingstandards.cpp.rules.resultofanassignmentoperatorshouldnotbeused.ResultOfAnAssignmentOperatorShouldNotBeUsed

class ResultOfAnAssignmentOperatorShouldNotBeUsedQuery extends ResultOfAnAssignmentOperatorShouldNotBeUsedSharedQuery
{
  ResultOfAnAssignmentOperatorShouldNotBeUsedQuery() {
    this = ImportMisra23Package::resultOfAnAssignmentOperatorShouldNotBeUsedQuery()
  }
}
