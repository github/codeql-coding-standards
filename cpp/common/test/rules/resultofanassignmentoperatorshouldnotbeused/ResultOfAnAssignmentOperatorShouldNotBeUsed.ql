// GENERATED FILE - DO NOT MODIFY
import codingstandards.cpp.rules.resultofanassignmentoperatorshouldnotbeused.ResultOfAnAssignmentOperatorShouldNotBeUsed

class TestFileQuery extends ResultOfAnAssignmentOperatorShouldNotBeUsedSharedQuery, TestQuery { }
