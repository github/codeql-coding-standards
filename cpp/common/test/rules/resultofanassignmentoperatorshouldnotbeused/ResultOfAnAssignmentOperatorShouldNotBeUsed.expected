| test.cpp:9:7:9:12 | ... = ... | Use of an assignment operator's result. |
| test.cpp:13:11:13:16 | ... = ... | Use of an assignment operator's result. |
| test.cpp:15:8:15:13 | ... = ... | Use of an assignment operator's result. |
